src/compiler/checker.ts

```diff
@@ -8228,6 +8358,27 @@ module ts {
                             break;
                     }
                 }
+
+                if (node.type) {
+                    if (languageVersion >= ScriptTarget.ES6 && isSyntacticallyValidGenerator(node)) {
+                        let returnType = getTypeFromTypeNode(node.type);
+                        if (returnType === voidType) {
+                            error(node.type, Diagnostics.A_generator_cannot_have_a_void_type_annotation);

```

Comment:
Discussed offline - @JsonFreeman suggested we can try using @ahejlsberg's suggested message as the head message for assignability.